### PR TITLE
Const fold optimizer

### PR DESCRIPTION
**Pull request by zhijxu-MS**

### CODE REVIEW

File: tf2onnx/optimizer/const_fold_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+from tf2onnx.optimizer.optimizer_base import GraphOptimizerBase
+from tf2onnx import utils
+
+# pylint: disable=logging-not-lazy,unused-argument,missing-docstring
+
+_func_map = {}
+
+
+def _register_func(onnx_op):
```
Comment (nbcsm) [inline tf2onnx/optimizer/const_fold_optimizer.py]: may need to consider domain and version as well, const fold logic may differ among different domains and versions.

Comment (zhijxu-MS) [reply tf2onnx/optimizer/const_fold_optimizer.py]: will add domain check.

File: tf2onnx/optimizer/const_fold_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+
+class ConstFoldOptimizer(GraphOptimizerBase):
+
+    def __init__(self, debug=False):
+        super(ConstFoldOptimizer, self).__init__("ConstFoldOptimizer", debug)
+        self._log = logging.getLogger("tf2onnx.optimizer.%s" % self._name)
+
+    def _optimize(self, graph):
+        return self._apply_optimization(graph, self._optimize_at_current_graph_level)
```
[the remark this answers is not in the document]
Comment (zhijxu-MS) [reply tf2onnx/optimizer/const_fold_optimizer.py]: already do it in function "optimize"

File: tf2onnx/optimizer/const_fold_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    @staticmethod
+    def _is_graph_output(node, graph):
+        node_out_set = set(node.output)
+        graph_out_set = set(graph.outputs)
+        return node_out_set.intersection(graph_out_set)
+
+    @staticmethod
+    def _replace_node_with_const(node, graph, val):
+        const_node = graph.make_const(utils.make_name("const_fold_opt"), val)
```
Comment (nbcsm) [inline tf2onnx/optimizer/const_fold_optimizer.py]: Set dtype and shape

File: tf2onnx/optimizer/transpose_optimizer.py [hunk 4]
```diff
@@ -41,6 +38,7 @@ class TransposeOptimizer(GraphOptimizerBase):
 
     def __init__(self, debug=False):
         super(TransposeOptimizer, self).__init__("TransposeOptimizer", debug)
+        self._log = logging.getLogger("tf2onnx.optimizer.%s" % self._name)
```
[the remark this answers is not in the document]
Comment (zhijxu-MS) [reply tf2onnx/optimizer/transpose_optimizer.py]: done.

File: tf2onnx/optimizer/const_fold_optimizer.py (hunk 3 above)
Comment (zhijxu-MS) [reply tf2onnx/optimizer/const_fold_optimizer.py]: done

File: tf2onnx/optimizer/optimizer_base.py [hunk 5]
```diff
@@ -13,10 +14,13 @@ class GraphOptimizerBase(object):
     def __init__(self, name, debug=False):
         self._debug = debug
         self._name = name
+        self._log = logging.getLogger("tf2onnx.optimizer.%s" % self._name)
 
     def optimize(self, graph):
         original_node_statistics = graph.dump_node_statistics()
         graph = self._optimize(graph)
+        graph.delete_unused_nodes(graph.outputs)
+        graph.topological_sort(graph.get_nodes())
```
Comment (nbcsm) [inline tf2onnx/optimizer/optimizer_base.py]: is topo-sort a common requirement for all optimizers?
i suggest let optimizer itself to call topo sort if topo order is required for that specific optimizer.

File: tf2onnx/optimizer/const_fold_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            for op in ops:
+                if self._should_skip(op):
+                    continue
+                if self._fold_node(op, graph):
+                    graph_changed = True
+        return graph
+
+    @staticmethod
+    def _should_skip(node):
+        # only support onnx official op for now, other op such as contrib op not supported for now
```
Comment (nbcsm) [inline tf2onnx/optimizer/const_fold_optimizer.py]: no "contrib op", just say "op in other domain"

File: tf2onnx/optimizer/const_fold_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        return False
+
+    def _fold_node(self, node, graph):
+        """ if node's input are all const and it's not graph's output then it can be fold.
+            if node can be fold True will be return indicating that graph is changed
+        """
+        if self._all_inputs_are_const(node.inputs) and not self._is_graph_output(node, graph):
+            process_func = _func_map.get(node.type, None)
+            if process_func:
+                const_val_after_trans = process_func(node, graph)
```
Comment (nbcsm) [inline tf2onnx/optimizer/const_fold_optimizer.py]: ```const_val_after_trans``` rename to ```const_outputs``` or ```outputs```?

File: tf2onnx/optimizer/const_fold_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                self._replace_node_with_const(node, graph, const_val_after_trans)
+                return True
+        self.log.warning("need to add function to fold op %s whose op_type is %s", node.name, node.type)
```
Comment (nbcsm) [inline tf2onnx/optimizer/const_fold_optimizer.py]: we already have too many warnings...
let's it be debug.

File: tf2onnx/optimizer/const_fold_optimizer.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+   if op's inputs are all const then do op computation when building the graph to improve performance
+   for example, input of transpose node is const then we can do transpose statically instead of at runtime
+"""
+
+from tf2onnx.optimizer.optimizer_base import GraphOptimizerBase
+from tf2onnx import utils
+
+# pylint: disable=logging-not-lazy,unused-argument,missing-docstring
+
+_func_map = {}
```
Comment (nbcsm) [inline tf2onnx/optimizer/const_fold_optimizer.py]: you may want to add a few comments for the expected signature of func.

File: tf2onnx/utils.py [hunk 10]
```diff
@@ -433,3 +433,9 @@ def create_vague_shape_like(shape):
 
 def get_onnx_version():
     return onnx.__version__
+
+
+def is_onnx_domain(domain):
+    if domain is None or domain == "":
+        return True
```
Comment (nbcsm) [inline tf2onnx/utils.py]: please resolve the conflict.

File: tf2onnx/optimizer/optimizer_base.py (hunk 5 above)
Comment (zhijxu-MS) [reply tf2onnx/optimizer/optimizer_base.py]: not necessary, I will remove it.

my initial thought is to make sure every optimizer ends with a legal onnx graph.

**Comment by zhijxu-MS** [comment]
@nbcsm add one commit(the last one) to enhance squeeze_op conversion log.

**Pull request merged**